### PR TITLE
 Create flags, extract uploadBytes and marshalGrid into functions

### PR DESCRIPTION
**Pull request by fejta**
Download config and update all groups specified in the config

Increase velocity of updating the entire config, each test group and each build of the group:
* Concurrently list builds
* Concurrently download json files and list artifacts within a build
* Concurrently download artifacts

Add flags to manually control concurrency

Extract messages from system-out, failure, skipped fields

Add some basic icons

### CODE REVIEW

File: testgrid/cmd/updater/main.go [hunk 1]
```diff
@@ -298,26 +328,38 @@ type Row struct {
...
+		}
+		r.Metrics = map[string]float64{
+			elapsedKey: float64(br.Finished - br.Started),
 		}
-		return state.Row_FAIL
 	case time.Now().Add(-24*time.Hour).Unix() > br.Started:
 		// Timed out
-		return state.Row_FAIL
+		r.Result = state.Row_FAIL
+		r.Message = "Never finished"
```
Comment (michelle192837) [inline testgrid/cmd/updater/main.go]: Maybe "Test didn't finish in 24 hours" or similar?

File: testgrid/cmd/updater/main.go [hunk 2]
```diff
@@ -771,6 +808,27 @@ func ReadBuild(build Build) (*Column, error) {
 	for t, rs := range rows {
 		br.Rows[t] = append(br.Rows[t], rs...)
 	}
+	if or.Result == state.Row_FAIL { // Ensure failing build has a failing row
+		ft := false
```
Comment (michelle192837) [inline testgrid/cmd/updater/main.go]: ft doesn't read clearly in this snippet (neither does 'if !ft' later)

File: testgrid/cmd/updater/main.go [hunk 3]
```diff
@@ -744,27 +782,28 @@ func main() {
...
+			log.Printf("Failed to list %s builds at %s: %v", o, tgPath, err)
 		}
 		close(builds)
 	}()
-	grid := ReadBuilds(*tg, builds, 3, Days(3))
+	grid := ReadBuilds(tg, builds, 3, Days(3))
 	buf, crc, err := marshalGrid(grid)
 	if err != nil {
-		log.Fatalf("Failed to marshal grid: %v", err)
+		return fmt.Errorf("failed to marhsal %s grid: %v", o, err)
```
Comment (michelle192837) [inline testgrid/cmd/updater/main.go]: nit: marshal

File: testgrid/cmd/updater/main.go [hunk 4]
```diff
@@ -631,82 +1059,133 @@ func Group(cfg config.Configuration, name string) (*config.TestGroup, bool) {
...
-	}()
-	grid := ReadBuilds(*tg, builds, 1000, Days(5))
-	log.Printf("Grid: %d %s", len(grid.Columns), grid.String())
+	}
+	log.Printf("WROTE: %s, %dx%d grid (%s, %d bytes)", tg.Name, len(grid.Columns), len(grid.Rows), tgp, len(buf))
+	return nil
+}
+
+// marhshalGrid serializes a state proto into zlib-compressed bytes and its crc32 checksum.
+func marshalGrid(grid state.Grid) ([]byte, uint32, error) {
```
Comment (rmmh) [inline testgrid/cmd/updater/main.go]: Why does this return a CRC? The only thing that cares about the CRC is uploadBytes, so it should continue to calculate it itself.

File: testgrid/cmd/updater/main.go [hunk 5]
```diff
@@ -42,13 +48,112 @@ import (
 	"vbom.ml/util/sortorder"
 )
 
+// options configures the updater
+type options struct {
+	config           gcsPath // gs://path/to/config/proto
+	creds            string  // TODO(fejta): implement
+	confirm          bool    // TODO(fejta): implement
+	group            string
+	groupConcurrency uint
```
Comment (rmmh) [inline testgrid/cmd/updater/main.go]: just use int for these two

File: testgrid/cmd/updater/main.go [hunk 6]
```diff
@@ -42,13 +48,112 @@ import (
...
+	buildConcurrency uint
+}
+
+// validate ensures sane options
+func (o *options) validate() error {
+	if o.config.String() == "" {
+		return o.config.Set("gs://fejternetes/config")
```
Comment (rmmh) [inline testgrid/cmd/updater/main.go]: why return early here?

File: testgrid/cmd/updater/main.go [hunk 7]
```diff
@@ -119,62 +225,108 @@ type JunitResult struct {
...
+	var msg string
+	switch {
+	case jr.Failure != nil && *jr.Failure != "":
+		msg = *jr.Failure
+	case jr.Skipped != nil && *jr.Skipped != "":
+		msg = *jr.Skipped
+	case jr.Output != nil && *jr.Output != "":
+		msg = *jr.Output
+	}
+	if l := len(msg); max == 0 || l <= max {
```
Comment (rmmh) [inline testgrid/cmd/updater/main.go]: `if max == 0 || len(msg) <= max {`

File: testgrid/cmd/updater/main.go [hunk 8]
```diff
@@ -119,62 +225,108 @@ type JunitResult struct {
...
+	return n, r
+}
+
+func unmarshalXML(buf []byte, i interface{}) error {
+	reader := bytes.NewReader(buf)
+	dec := xml.NewDecoder(reader)
+	dec.CharsetReader = func(charset string, input io.Reader) (io.Reader, error) {
+		switch charset {
+		case "UTF-8", "utf8", "":
+			// utf8 is not recognized by golang, but our coalesce.py writes a utf8 doc, which python accepts.
```
Comment (rmmh) [inline testgrid/cmd/updater/main.go]: We should *probably* fix this!

File: testgrid/cmd/updater/main.go (hunk 5 above)
Comment (fejta) [reply testgrid/cmd/updater/main.go]: Done

File: testgrid/cmd/updater/main.go (hunk 6 above)
Comment (fejta) [reply testgrid/cmd/updater/main.go]: Done

File: testgrid/cmd/updater/main.go (hunk 7 above)
Comment (fejta) [reply testgrid/cmd/updater/main.go]: `l` is used in the `else` statement (`msg[l-h-1:]`)

File: testgrid/cmd/updater/main.go (hunk 8 above)
Comment (fejta) [reply testgrid/cmd/updater/main.go]: Ack

File: testgrid/cmd/updater/main.go (hunk 1 above)
Comment (fejta) [reply testgrid/cmd/updater/main.go]: Done

File: testgrid/cmd/updater/main.go (hunk 4 above)
Comment (fejta) [reply testgrid/cmd/updater/main.go]: Done

**Comment by fejta** [comment]
PTAL

/hold
(want to squash before merging)

**Comment by fejta** [comment]
PTAL?

**Pull request merged**